File: frontend/src/app/shared/services/booking.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable, forkJoin, throwError, of } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { ApiService } from '../../core/services/api.service';
import {
  ShowtimeSeatDto,
  BookingRequest,
  BookingResponse,
  SeatLockRequest,
} from '../../core/models/backend-dtos';

@Injectable({
  providedIn: 'root',
})
export class BookingService {
  private apiService = inject(ApiService);

  /**
   * Get all seats for a showtime with their status
   */
  getShowtimeSeats(showtimeId: number): Observable<ShowtimeSeatDto[]> {
    return this.apiService.get<ShowtimeSeatDto[]>(`/showtimes/${showtimeId}/seats`).pipe(
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }

  /**
   * PHASE 1: Lock a single seat
   */
  lockSeat(showtimeSeatId: number): Observable<void> {
    const request: SeatLockRequest = { showtimeSeatId };
    return this.apiService.post<void>('/seats/lock', request).pipe(
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }

  /**
   * PHASE 1: Lock multiple seats sequentially or in parallel
   */
  lockSeats(showtimeSeatIds: number[]): Observable<void[]> {
    if (showtimeSeatIds.length === 0) {
      return throwError(() => new Error('No seats selected'));
    }

    // Create array of lock requests
    const lockRequests: Observable<void>[] = showtimeSeatIds.map((seatId) => this.lockSeat(seatId));

    // Execute all lock requests in parallel
    return forkJoin(lockRequests).pipe(
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }

  /**
   * PHASE 2: Create booking with locked seats
   */
  createBooking(showtimeSeatIds: number[]): Observable<BookingResponse> {
    const request: BookingRequest = { showtimeSeatIds };

    return this.apiService.post<BookingResponse>('/bookings', request).pipe(
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }

  /**
   * Get booking by UUID (for ticket page)
   */
  getBookingByUuid(uuid: string): Observable<BookingResponse> {
    return this.apiService.get<BookingResponse>(`/bookings/${uuid}`).pipe(
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }

  /**
   * Get all bookings for current user
   */
  getMyBookings(): Observable<BookingResponse[]> {
    return this.apiService.get<BookingResponse[]>('/bookings/my-bookings').pipe(
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }

  /**
   * Cancel a booking by UUID (not used in user flow, kept for compatibility)
   */
  cancelBooking(uuid: string): Observable<void> {
    return this.apiService.delete<void>(`/bookings/${uuid}`).pipe(
      catchError((error) => {
        return throwError(() => error);
      })
    );
  }
}

File: backend/src/main/java/com/ohnomovies/backend/service/booking/BookingService.java
package com.ohnomovies.backend.service.booking;

import java.util.List;
import java.util.UUID;

import org.springframework.transaction.annotation.Transactional;

import com.ohnomovies.backend.dao.dto.booking.BookingRequest;
import com.ohnomovies.backend.dao.dto.booking.BookingResponse;
import com.ohnomovies.backend.dao.dto.booking.BookingVerificationResponse;
import com.ohnomovies.backend.model.entity.User;

public interface BookingService {
    @Transactional
    void lockSeat(Long showtimeSeatId, User user);

    @Transactional
    BookingResponse createBooking(BookingRequest request, User user);

    @Transactional(readOnly = true)
    List<BookingResponse> getMyBookings(User user);

    @Transactional(readOnly = true)
    BookingResponse getUserBookingByUuid(UUID uuid, User user);

    @Transactional(readOnly = true)
    List<BookingResponse> getAllBookings();

    @Transactional(readOnly = true)
    BookingResponse getBookingByUuid(UUID uuid);

    @Transactional
    BookingVerificationResponse verifyBooking(UUID uuid);

    @Transactional
    void confirmPaymentAndBooking(UUID bookingUuid, String paymentIntentId, String paymentMethod);

    @Transactional
    void cancelBooking(UUID uuid);
}
